### PR TITLE
Nullable type feature

### PR DESCRIPTION
**Pull request by vietj**
Annotate types to declare them nullable or not, for instance:

```
@Nullable Cookie getCookie(String name);
void close(@Nullable Handler<Void> closeHandler);
void publish(String address, @Nullable Object msg);
@Nullable V put(K key, V value);
@Nullable JsonObject config();
etc...
```

### CODE REVIEW

File: src/main/asciidoc/java/index.adoc [hunk 1]
```diff
@@ -320,6 +320,74 @@ void add(double x, double y);
...
+- some shims can leverage this information to provide a better API, for instance an `Optional<String>` Java type or the
+`String?` in Ceylon, etc...
+
+Codegen provides the `link:../../apidocs/io/vertx/codegen/annotations/Nullable.html[@Nullable]` annotations for annotating types.
+
+Method return type can be `link:../../apidocs/io/vertx/codegen/annotations/Nullable.html[@Nullable]`:
+
+[source,java]
+----
+Nullable String getAttribute(String name);
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: @ missing

File: src/main/asciidoc/java/index.adoc [hunk 2]
```diff
@@ -320,6 +320,74 @@ void add(double x, double y);
...
+----
+
+As well as method parameter type:
+
+[source,java]
+----
+void close(Nullable Handler<Void> closeHandler);
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: @ missing

File: src/main/asciidoc/java/index.adoc [hunk 3]
```diff
@@ -320,6 +320,74 @@ void add(double x, double y);
...
+----
+
+WARNING: type validation is a non goal for this feature, its purpose is to give hints to the shim
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: is not the goal of this feature

File: src/main/asciidoc/java/index.adoc [hunk 4]
```diff
@@ -320,6 +320,74 @@ void add(double x, double y);
...
+. `io.vertx.core.Handler<io.vertx.core.AsyncResult>` type argument can be `link:../../apidocs/io/vertx/codegen/annotations/Nullable.html[@Nullable]`
+but not for `java.lang.Void`
+. the `java.lang.Object` type is always nullable
+. the `<T>` in `<T>` parameter/return, `Handler<T>` or `Handler<AsyncResult<T>>` is always nullable
+. a method overriding another method `inherits` the `link:../../apidocs/io/vertx/codegen/annotations/Nullable.html[@Nullable]` usage of the overriden method
+. a method overriding another method cannot declare `link:../../apidocs/io/vertx/codegen/annotations/Nullable.html[@Nullable]` in its types
+
+In addition these rules apply to `link:../../apidocs/io/vertx/codegen/annotations/Nullable.html[@Nullable]` type arguments:
+
+. methods cannot declare generic api types with nullable type arguments, e.g `<T> void method(GenericApi<Nullable T> api)` is not permitted
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: @ missing

File: src/main/asciidoc/java/index.adoc [hunk 5]
```diff
@@ -320,6 +320,74 @@ void add(double x, double y);
...
+. methods can declare nullable collection, e.g `void method(List<Nullable String> list)` is allowed
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: @ missing

**Comment by cescoffier** [comment]
looks good except the @ missing in the documentation.

I' tried to think to edge case not covered by the TCK, I cannot see anything. 

One question:

Has @Nullable JsonArray the same semantic as @Nullable List<X> ?

**Comment by vietj** [comment]
`@Nullable JsonArray` means the JsonArray param or return can be null
`@Nullable List<X>` : has the same semantic

**Comment by cescoffier** [comment]
So both can contains null values.

**Comment by vietj** [comment]
no, then it would be `List<@Nullable String>` that is equivalent to `JsonArray`.

**Pull request merged**